apex_launchtest_ros/examples/talker_listener.test.py

```diff
@@ -0,0 +1,189 @@
+import launch_ros.actions
+import rclpy
+import rclpy.context
+import rclpy.executors
+import std_msgs.msg
+
+import apex_launchtest_ros
+from apex_launchtest.util import NoMatchingProcessException
+
+
+def generate_test_description(ready_fn):
+    # Necessary to get real-time stdout from python processes:
+    proc_env = os.environ.copy()
+    proc_env["PYTHONUNBUFFERED"] = "1"
+
+    # Normally, talker publishes on the 'chatter' topic and listener listens on the
+    # 'chatter' topic, but we want to show how to use remappings to munge the data so we
+    # will remap these topics when we launch the nodes and insert our own node that can
+    # change the data as it passes through
+    talker_node = launch_ros.actions.Node(
+        package="demo_nodes_py",
+        node_executable="talker",
+        env=proc_env,
+        remappings=[("chatter", "talker_chatter")],
+    )
+
+    listener_node = launch_ros.actions.Node(
+        package="demo_nodes_py",
+        node_executable="listener",
+        env=proc_env,
+    )
+
+    return (
+        launch.LaunchDescription([
+            talker_node,
+            listener_node,
+            # Start tests right away - no need to wait for anything
+            launch.actions.OpaqueFunction(function=lambda context: ready_fn()),
+        ]),
+        {
+            "talker": talker_node,
+            "listener": listener_node,
+        }
+    )
+
+
+class TestTalkerListenerLink(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls, proc_output, listener):
+        cls.context = rclpy.context.Context()
+        rclpy.init(context=cls.context)
+        cls.node = rclpy.create_node("test_node", context=cls.context)
+
+        # The demo node listener has no synchronization to indicate when it's ready to start
+        # receiving messages on the /chatter topic.  This plumb_listener method will attempt
+        # to publish for a few seconds until it sees output
+        publisher = cls.node.create_publisher(
+            std_msgs.msg.String,
+            "chatter"
+        )
+        msg = std_msgs.msg.String()
+        msg.data = "test message {}".format(uuid.uuid4())
+        for _ in range(5):
+            try:
+                publisher.publish(msg)
+                proc_output.assertWaitFor(
+                    msg=msg.data,
+                    process=listener,
+                    timeout=1.0
+                )
+            except AssertionError:
+                continue
+            except NoMatchingProcessException:
+                continue
+            else:
+                return
+        else:
+            assert False, "Failed to plumb chatter topic to listener process"
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.node.destroy_node()
+
+    def spin_rclpy(self, timeout_sec):
+        executor = rclpy.executors.SingleThreadedExecutor(context=self.context)
+        executor.add_node(self.node)
+        try:
+            executor.spin_once(timeout_sec=timeout_sec)
```

Comment:
If the launch description has any lifecycle nodes in it, it will also require a [ROSSpecificLaunchStartup](https://github.com/ros2/launch/blob/master/launch_ros/launch_ros/default_launch_description.py#L70) to the launch description.  This starts a thread that [will spin in the background](https://github.com/ros2/launch/blob/master/launch_ros/launch_ros/default_launch_description.py#L92) and will conflict with this spin until [this issue is resolved](https://github.com/ros2/rclpy/issues/192)

This example doesn't use lifecycle nodes, so it'll launch without the ROSSpecificLaunchStartup